rust/agama-dbus-server/src/locale.rs

```diff
@@ -0,0 +1,199 @@
+use crate::error::Error;
+use anyhow::Context;
+use std::process::Command;
+use zbus::{dbus_interface, Connection, ConnectionBuilder};
+
+pub struct Locale {
+    locales: Vec<String>,
+    keymap: String,
+    timezone_id: String,
+    supported_locales: Vec<String>,
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Locale1")]
+impl Locale {
+    // Can be `async` as well.
+    /// get labels for given locale. The first pair is english language and territory
+    /// and second one is localized one to target language from locale.
+    ///
+    /// Note: check how often it is used and if often, it can be easily cached
+    fn labels_for_locales(&self) -> Result<Vec<((String, String), (String, String))>, Error> {
+        const DEFAULT_LANG: &str = "en";
+        let mut res = Vec::with_capacity(self.supported_locales.len());
+        let languages = agama_locale_data::get_languages()?;
+        let territories = agama_locale_data::get_territories()?;
+        for locale in self.supported_locales.as_slice() {
+            let (loc_language, loc_territory) = agama_locale_data::parse_locale(locale.as_str())?;
+
+            let language = languages.find_by_id(loc_language)
+                .context("language for passed locale not found")?;
+            let territory = territories.find_by_id(loc_territory)
+                .context("territory for passed locale not found")?;
+
+            let default_ret = (
+                language
+                    .names
+                    .name_for(DEFAULT_LANG)
+                    .context("missing default translation for language")?,
+                territory
+                    .names
+                    .name_for(DEFAULT_LANG)
+                    .context("missing default translation for territory")?,
+            );
+            let localized_ret = (
+                language
+                    .names
+                    .name_for(language.id.as_str())
+                    .context("missing native label for language")?,
+                territory
+                    .names
+                    .name_for(language.id.as_str())
+                    .context("missing native label for territory")?,
+            );
+            res.push((default_ret, localized_ret));
+        }
+
+        Ok(res)
+    }
+
+    #[dbus_interface(property)]
+    fn locales(&self) -> Vec<String> {
+        return self.locales.to_owned();
+    }
+
+    #[dbus_interface(property)]
+    fn set_locales(&mut self, locales: Vec<String>) -> Result<(), zbus::fdo::Error> {
```

Comment:
Oh, I overlooked this one during the review. If you prefer, this is equivalent to `zbus::fdo::Result<()>`.